azext_iot/iothub/providers/state.py

```diff
@@ -0,0 +1,692 @@
+
+            logger.info("Saved state of IoT Hub '{}' to {}".format(self.hub_name, state_file))
+
+        except FileNotFoundError:
+            raise FileOperationError(f'File {state_file} does not exist.')
+
+    def upload_state(self, state_file: str, replace: Optional[bool] = None, hub_aspects: List[str] = None):
+        '''Uses device info from file to recreate the hub state'''
+        if not hub_aspects:
+            hub_aspects = HubAspects.list()
+
+        self.delete_aspects(replace, hub_aspects)
+
+        try:
+            with open(state_file, 'r', encoding='utf-8') as f:
+                hub_state = json.load(f)
+
+            self.upload_hub_from_dict(hub_state, hub_aspects)
+            logger.info("Uploaded state from '{}' to IoT Hub '{}'".format(state_file, self.hub_name))
+
+        except FileNotFoundError:
+            raise FileOperationError(f'File {state_file} does not exist.')
+
+    def migrate_state(
+        self,
+        orig_hub: Optional[str] = None,
+        orig_rg: Optional[str] = None,
+        orig_hub_login: Optional[str] = None,
+        replace: Optional[bool] = False,
+        hub_aspects: List[str] = None
+    ):
+        '''Migrates state from original hub to destination hub.'''
+        orig_hub_target = self.discovery.get_target(
+            resource_name=orig_hub,
+            resource_group_name=orig_rg,
+            login=orig_hub_login,
+            auth_type=self.auth_type
+        )
+
+        if not hub_aspects:
+            hub_aspects = HubAspects.list()
+
+        # Command modifies hub_aspect - make copy so we can reuse for upload
+        hub_state = self.process_hub_to_dict(orig_hub_target, hub_aspects[:])
+        self.delete_aspects(replace, hub_aspects)
+        self.upload_hub_from_dict(hub_state, hub_aspects)
+
+        logger.info("Migrated state from IoT Hub '{}' to {}".format(orig_hub, self.hub_name))
+
+    def delete_aspects(self, replace, hub_aspects=[]):
+        """
+        Delete all necessary hub aspects if the hub exists.
+
+        Delete dataplane aspects only present in hub_aspects.
+        """
+        if self.target and replace:
+            if HubAspects.Configurations.value in hub_aspects:
+                self.delete_all_configs()
+            if HubAspects.Devices.value in hub_aspects:
+                self.delete_all_devices()
+            if HubAspects.Arm.value in hub_aspects:
+                cert_client = self.get_client().certificates
+                # serialize strips name and etag - use as_dict instead
+                certificates = cert_client.list_by_iot_hub(self.rg, self.hub_name).as_dict()
+
+                for cert in tqdm(certificates["value"], desc="Deleting certificates from destination hub", ascii=" #"):
+                    cert_client.delete(self.rg, self.hub_name, cert["name"], cert["etag"])
+
+    def process_hub_to_dict(self, target: dict, hub_aspects: list = []) -> dict:
```

Comment:
Please avoid using an empty list as the default parameter value since list is a mutable type. Instead it is better for the parameter to have the default value of `None` and the function body can test for None and if so assign hub_aspects to an empty list.

Reply:
See this reference for more details https://docs.python.org/3/reference/compound_stmts.html#function-definitions

Reply:
removing the optional param since the only places it is called already defines the list.